### PR TITLE
Add initial tracing architecture

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Trace transfer events in go-data-transfer, link them to graphsync requests

# Implementation

- My approach to tracing spans is:
   - a data transfer span begins whenever we get the first event related to a data transfer channel ID
      - typical ways this might happen are:
         - OpenPushDataTransferChannel called
         - OpenPullDataTransferChannel called
         - RestartDataTransferChannel called (but not always -- would only start a span on a node restart, as opposed to a channel monitor restart)
         - Incoming graphsync request with data transfer message queued on channel that currently has no state
         - Incoming message on libp2p protocol on channel that currently has no state
      - the basic approach is "create if not exist"
   - it ends when the associated channel closes either in success or error, or the program shuts down
- My thought is it isn't so important to get exact starts and stops as much as it is to capture that it DOES start and stop, so that everything below, including graphsync traces will be situated correctly

# To Do

- Add tests! 
- Possibly some spans in the transport for graphsync

### CODE REVIEW

File: impl/receiver.go [hunk 1]
```diff
@@ -91,6 +105,18 @@ func (r *receiver) receiveResponse(
 	sender peer.ID,
 	incoming datatransfer.Response) error {
 	chid := datatransfer.ChannelID{Initiator: r.manager.peerID, Responder: sender, ID: incoming.TransferID()}
+	ctx, _ = r.manager.spansIndex.SpanForChannel(ctx, chid)
+	ctx, span := otel.Tracer("data-transfer").Start(ctx, "receiveRequest", trace.WithAttributes(
```
Comment (dirkmc) [inline impl/receiver.go:109]:
```suggestion
	ctx, span := otel.Tracer("data-transfer").Start(ctx, "receiveResponse", trace.WithAttributes(
```

File: tracing/tracing.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (si *SpansIndex) SpanForChannel(ctx context.Context, chid datatransfer.ChannelID) (context.Context, trace.Span) {
+	si.spansLk.RLock()
+	span, ok := si.spans[chid]
+	if ok {
+		si.spansLk.RUnlock()
+		return trace.ContextWithSpan(ctx, span), span
+	}
+	si.spansLk.RUnlock()
```
Comment (dirkmc) [inline tracing/tracing.go:32]:
```suggestion
	si.spansLk.RUnlock()
	if ok {
		return trace.ContextWithSpan(ctx, span), span
	}
```

File: tracing/tracing.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	si.spansLk.Lock()
+	defer si.spansLk.Unlock()
+	// need to recheck under the write lock
+	span, ok = si.spans[chid]
+	if ok {
+		return trace.ContextWithSpan(ctx, span), span
+	}
+	return otel.Tracer("data-transfer").Start(ctx, "transfer", trace.WithAttributes(
```
Comment (dirkmc) [inline tracing/tracing.go:40]: Should the span be inserted into the map here?

File: transport.go [hunk 4]
```diff
@@ -69,6 +69,8 @@ type EventsHandler interface {
 	// OnReceiveDataError is called when a network error occurs receiving data
 	// at the transport layer
 	OnReceiveDataError(chid ChannelID, err error) error
+
+	OnContextAugment(chid ChannelID) func(context.Context) context.Context
```
Comment (dirkmc) [inline transport.go:73]: Suggest adding a comment to explain what this is

File: transport/graphsync/graphsync.go [hunk 5]
```diff
@@ -1103,6 +1107,35 @@ func (c *dtChannel) gsReqOpened(gsKey graphsyncKey, hookActions graphsync.Outgoi
 	c.opened <- gsKey
 }
 
+func (c *dtChannel) gsDataRequestQueued(gsKey graphsyncKey, hookActions graphsync.IncomingRequestHookActions) {
```
Comment (dirkmc) [inline transport/graphsync/graphsync.go:1110]: Is this method being called from somewhere?

Comment (hannahhoward) [reply transport/graphsync/graphsync.go:1110]: ha oops forgot i added that

File: impl/impl.go [hunk 6]
```diff
@@ -209,6 +217,7 @@ func (m *manager) OpenPushDataChannel(ctx context.Context, requestTo peer.ID, vo
 	monitoredChan := m.channelMonitor.AddPushChannel(chid)
 	if err := m.dataTransferNetwork.SendMessage(ctx, requestTo, req); err != nil {
 		err = fmt.Errorf("Unable to send request: %w", err)
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:220]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

Comment (rvagg) [reply impl/impl.go:220]: assuming these are fatal errors?

File: impl/impl.go [hunk 7]
```diff
@@ -247,14 +257,14 @@ func (m *manager) OpenPullDataChannel(ctx context.Context, requestTo peer.ID, vo
 	monitoredChan := m.channelMonitor.AddPullChannel(chid)
 	if err := m.transport.OpenChannel(ctx, requestTo, chid, cidlink.Link{Cid: baseCid}, selector, nil, req); err != nil {
 		err = fmt.Errorf("Unable to send request: %w", err)
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:260]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: impl/impl.go [hunk 8]
```diff
@@ -266,16 +276,26 @@ func (m *manager) SendVoucher(ctx context.Context, channelID datatransfer.Channe
...
+	ctx, _ = m.spansIndex.SpanForChannel(ctx, channelID)
+	ctx, span := otel.Tracer("data-transfer").Start(ctx, "sendVoucher", trace.WithAttributes(
+		attribute.String("channelID", channelID.String()),
+		attribute.String("voucherType", string(voucher.Type())),
+	))
+	defer span.End()
 	if channelID.Initiator != m.peerID {
-		return errors.New("cannot send voucher for request we did not initiate")
+		err := errors.New("cannot send voucher for request we did not initiate")
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:287]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: impl/impl.go [hunk 9]
```diff
@@ -266,16 +276,26 @@ func (m *manager) SendVoucher(ctx context.Context, channelID datatransfer.Channe
...
+		return err
 	}
 	updateRequest, err := message.VoucherRequest(channelID.ID, voucher.Type(), voucher)
 	if err != nil {
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:292]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: impl/impl.go [hunk 10]
```diff
@@ -266,16 +276,26 @@ func (m *manager) SendVoucher(ctx context.Context, channelID datatransfer.Channe
...
 		return err
 	}
 	if err := m.dataTransferNetwork.SendMessage(ctx, chst.OtherPeer(), updateRequest); err != nil {
 		err = fmt.Errorf("Unable to send request: %w", err)
 		_ = m.OnRequestDisconnected(channelID, err)
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:298]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: impl/impl.go [hunk 11]
```diff
@@ -289,10 +309,15 @@ func (m *manager) CloseDataTransferChannel(ctx context.Context, chid datatransfe
...
-
+	ctx, _ = m.spansIndex.SpanForChannel(ctx, chid)
+	ctx, span := otel.Tracer("data-transfer").Start(ctx, "closeChannel", trace.WithAttributes(
+		attribute.String("channelID", chid.String()),
+	))
+	defer span.End()
 	// Close the channel on the local transport
 	err = m.transport.CloseChannel(ctx, chid)
 	if err != nil {
+		span.RecordError(err)
```
Comment (rvagg) [inline impl/impl.go:320]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: network/libp2p_impl.go [hunk 12]
```diff
@@ -160,19 +163,36 @@ func (dtnet *libp2pDataTransferNetwork) SendMessage(
...
+		attribute.Bool("isRestart", outgoing.IsRestart()),
+		attribute.Bool("isUpdate", outgoing.IsUpdate()),
+		attribute.Bool("isCancel", outgoing.IsCancel()),
+		attribute.Bool("isPaused", outgoing.IsPaused()),
+	))
+
+	defer span.End()
 	s, err := dtnet.openStream(ctx, p, dtnet.dtProtocols...)
 	if err != nil {
+		span.RecordError(err)
```
Comment (rvagg) [inline network/libp2p_impl.go:180]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: network/libp2p_impl.go [hunk 13]
```diff
@@ -160,19 +163,36 @@ func (dtnet *libp2pDataTransferNetwork) SendMessage(
...
 		return err
 	}
 
 	outgoing, err = outgoing.MessageForProtocol(s.Protocol())
 	if err != nil {
-		return xerrors.Errorf("failed to convert message for protocol: %w", err)
+		err = xerrors.Errorf("failed to convert message for protocol: %w", err)
+		span.RecordError(err)
```
Comment (rvagg) [inline network/libp2p_impl.go:187]:
```suggestion
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
```

File: network/libp2p_impl.go [hunk 14]
```diff
@@ -160,19 +163,36 @@ func (dtnet *libp2pDataTransferNetwork) SendMessage(
...
 	}
 
 	if err = dtnet.msgToStream(ctx, s, outgoing); err != nil {
+		span.RecordError(err)
 		if err2 := s.Reset(); err2 != nil {
 			log.Error(err)
+			span.RecordError(err2)
 			return err2
 		}
 		return err
```
Comment (rvagg) [inline network/libp2p_impl.go:198]:
```suggestion
		if err2 := s.Reset(); err2 != nil {
			log.Error(err)
			span.RecordError(err2)
			span.SetStatus(codes.Error, err2.Error())
			return err2
		}
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return err
```

Comment (rvagg) [reply network/libp2p_impl.go:198]: not sure about that shuffle tbh, maybe recording both errors is the right approach, but the `SetStatus()` should probably record just the fatal one for each branch

Comment (hannahhoward) [reply network/libp2p_impl.go:200]: genuinely, I'm not sure what the right path is!

**Pull request merged**